debugger/gdb-sgx-plugin/mman.py

```diff
@@ -0,0 +1,506 @@
+    SIZEOF_PREV=8
+
+    OFFSETOF_NEXT=8
+    SIZEOF_NEXT=8
+
+    OFFSETOF_PID=16
+    SIZEOF_PID=4
+
+    OFFSETOF_NMAPS=20
+    SIZEOF_NMAPS=24
+
+    def __init__(self, addr):
+        if addr:
+            self.prev = read_int_from_memory(addr, self.SIZEOF_PREV)
+        else:
+            return
+
+        self.next = read_int_from_memory(addr + self.OFFSETOF_NEXT, self.SIZEOF_NEXT)
+        self.pid = read_int_from_memory(addr + self.OFFSETOF_PID, self.SIZEOF_PID)
+        self.nmaps = read_int_from_memory(addr + self.OFFSETOF_NMAPS, self.SIZEOF_NMAPS)
+
+SHMEM_TYPE_ANON = 1
+SHMEM_TYPE_REG_FILE = 2
+SHMEM_TYPE_POSIX_SHM = 3
+class shared_mapping_t:
+    OFFSETOF_PREV=0
+    SIZEOF_PREV=8
+
+    OFFSETOF_NEXT=8
+    SIZEOF_NEXT=8
+
+    OFFSETOF_SHARERS_HEAD=16
+    SIZEOF_SHARERS_HEAD=8
+
+    OFFSETOF_SHARERS_TAIL=24
+    SIZEOF_SHARERS_TAIL=8
+
+    OFFSETOF_SHARERS_SIZE=32
+    SIZEOF_SHARERS_SIZE=8
+
+    OFFSETOF_START_ADDR=40
+    SIZEOF_START_ADDR=8
+
+    OFFSETOF_LENGTH=48
+    SIZEOF_LENGTH=8
+
+    OFFSETOF_FILE_SIZE=56
+    SIZEOF_FILE_SIZE=8
+
+    OFFSETOF_FILE_HANDLE=64
+    SIZEOF_FILE_HANDLE=8
+
+    OFFSETOF_OFFSET=72
+    SIZEOF_OFFSET=8
+
+    OFFSETOF_TYPE=80
+    SIZEOF_TYPE=4
+
+    def __init__(self, addr):
+        if addr:
+            self.prev = read_int_from_memory(addr + self.OFFSETOF_PREV, self.SIZEOF_PREV)
+        else:
+            return
+
+        self.next = read_int_from_memory(addr + self.OFFSETOF_NEXT, self.SIZEOF_NEXT)
+        self.sharers_head = read_int_from_memory(addr + self.OFFSETOF_SHARERS_HEAD, self.SIZEOF_SHARERS_HEAD)
+        self.sharers_tail = read_int_from_memory(addr + self.OFFSETOF_SHARERS_TAIL, self.SIZEOF_SHARERS_TAIL)
+        self.sharers_size = read_int_from_memory(addr + self.OFFSETOF_SHARERS_SIZE, self.SIZEOF_SHARERS_SIZE)
+        self.start_addr = read_int_from_memory(addr + self.OFFSETOF_START_ADDR, self.SIZEOF_START_ADDR)
+        self.length = read_int_from_memory(addr + self.OFFSETOF_LENGTH, self.SIZEOF_LENGTH)
+        self.file_size = read_int_from_memory(addr + self.OFFSETOF_FILE_SIZE, self.SIZEOF_FILE_SIZE)
+        self.file_handle = read_int_from_memory(addr + self.OFFSETOF_FILE_HANDLE, self.SIZEOF_FILE_HANDLE)
+        self.offset = read_int_from_memory(addr + self.OFFSETOF_OFFSET, self.SIZEOF_OFFSET)
+        self.type = read_int_from_memory(addr + self.OFFSETOF_TYPE, self.SIZEOF_TYPE)
+
+class MystShmfsInitBreakpoint(gdb.Breakpoint):
+    def __init__(self):
+        super(MystShmfsInitBreakpoint, self).__init__('myst_shmfs_setup_hook', internal=False)
+
+    def stop(self):
+        mman_tracker.shared_mappings = int(gdb.parse_and_eval('(uint64_t)&_shared_mappings'))
+
+class MystMmanInitBreakpoint(gdb.Breakpoint):
+    def __init__(self):
+        super(MystMmanInitBreakpoint, self).__init__('myst_mman_init_debug_hook', internal=False)
+
+    def stop(self):
+        mman_tracker.mman_base = int(gdb.parse_and_eval('(uint64_t)_mman->base'))
```

Comment:
Ideally we'd want to rely on offsets instead of field names. Field names require debugging information.

Also ideally, we'd use register instead of variable name _mman.

Reply:
By using register are you suggesting passing a pointer to `_mman` to the debugger hook function, and then reading off `rdi`?